### PR TITLE
PLF-8344: Restrict actions for specific folders

### PR DESCRIPTION
**Pull request by wkhessairi**
Restrict "rename", "delete" and "cut" actions for specific folders:  Documents, Favorites, Pictures, Music, videos & Public.
This is performed by adding an extension filter "IsNotSpecificFolderNodeFilter" that does not accept those specific folders.
A hide of actions is also performed on views (for multi folders selection and different views). Adding a new attribute "isSpecificFolder" for nodes that are specific folders. And then hide the corresponding actions on simple, file and list views by adding css display none

### CODE REVIEW

File: core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java [hunk 1]
```diff
@@ -89,6 +90,9 @@
 
   public static final String PUBLIC = "Public";
 
+  public static final String[] SPECIFIC_FOLDERS = { NodetypeConstant.EXO_DOCUMENTFOLDER, NodetypeConstant.EXO_FAVOURITE_FOLDER,
+      NodetypeConstant.EXO_PICTUREFOLDER, NodetypeConstant.EXO_MUSICFOLDER, NodetypeConstant.EXO_VIDEOFOLDER };
```
Comment (boubaker) [inline core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java]: We should include hidden folders here as well, please see:
https://github.com/exodev/platform/blob/develop/extension/webapp/src/main/webapp/WEB-INF/conf/jcr/jcr-configuration.xml#L286
* exo:searchFolder
* exo:folksonomyFolder

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/filter/IsNotSpecificFolderNodeFilter.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package org.exoplatform.ecm.webui.component.explorer.control.filter;
+
+import java.util.Map;
+import javax.jcr.Node;
+
+import org.exoplatform.services.cms.impl.Utils;
+import org.exoplatform.webui.ext.filter.UIExtensionAbstractFilter;
+import org.exoplatform.webui.ext.filter.UIExtensionFilterType;
+
+public class IsNotSpecificFolderNodeFilter extends UIExtensionAbstractFilter {
```
Comment (boubaker) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/filter/IsNotSpecificFolderNodeFilter.java]: please add javadoc to describe class main role

File: core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java [hunk 3]
```diff
@@ -969,4 +973,28 @@ public static boolean isFolder(Node node) throws RepositoryException  {
         || node.isNodeType(NodetypeConstant.NT_UNSTRUCTURED);
  }
 
+ public static boolean isPersonalSpecificFolder(Node node) throws Exception {
```
Comment (boubaker) [inline core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java]: Rename this method to be make it more comprehensive, like `isPersonalDefaultFolder` and please add javadoc in both newly added methods

File: apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js [hunk 4]
```diff
@@ -648,14 +648,16 @@
...
   	  
       for (var i in Self.itemsSelected) {
         if (Array.prototype[i]) continue;
         // check if a node is exo:action or not to show nothing on action bar.
   	  	if (	Self.itemsSelected[i].getAttribute('isExoAction') == "true") {
   		  checkExoActionNode = true;
   		  break;
-  	  	}  
+  	  	}
+        if (Self.itemsSelected[i].getAttribute('isSpecificFolder') == "true") checkSpecificFolder = true;
```
Comment (boubaker) [inline apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js]: Please use braces after if and return to new line

File: apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js [hunk 5]
```diff
@@ -648,14 +648,16 @@
       var checkLinkAndTargetInTrash = false;
       var checkExoActionNode = false;
       var checkInStatus = false;
+      var checkSpecificFolder = false;
```
Comment (boubaker) [inline apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js]: Please rename variable to `isUserPrivateFolder`

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/filter/IsNotSpecificFolderNodeFilter.java (hunk 2 above)
Comment (wkhessairi) [reply core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/control/filter/IsNotSpecificFolderNodeFilter.java]: done

File: core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java (hunk 1 above)
Comment (wkhessairi) [reply core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java]: done

File: apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js (hunk 4 above)
Comment (wkhessairi) [reply apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js]: done

File: apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js (hunk 5 above)
Comment (wkhessairi) [reply apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js]: done

File: apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js [hunk 6]
```diff
@@ -648,14 +648,18 @@
...
   	  
       for (var i in Self.itemsSelected) {
         if (Array.prototype[i]) continue;
         // check if a node is exo:action or not to show nothing on action bar.
   	  	if (	Self.itemsSelected[i].getAttribute('isExoAction') == "true") {
   		  checkExoActionNode = true;
   		  break;
-  	  	}  
+  	  	}
+  	  	if (Self.itemsSelected[i].getAttribute('isSpecificFolder') == "true") {
```
Comment (boubaker) [inline apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js]: you can make it like this:
`isUserPrivateFolder = isUserPrivateFolder || Self.itemsSelected[i].getAttribute('isSpecificFolder') == "true"`

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/UIDocumentInfo.java [hunk 7]
```diff
@@ -1850,6 +1850,8 @@ private String getNodeAttributeInCommon(Node node) throws Exception {
     builder.append(" isAbleToRestore='" + Utils.isAbleToRestore(node) + "' ");
     builder.append(" isExoAction='" + (Utils.EXO_ACTIONS.equals(node.getName()) && Utils.isInTrash(node)) + "' ");
     builder.append(" isCheckedIn='" + !node.isCheckedOut() + "' ");
+    builder.append(" isSpecificFolder='" + (org.exoplatform.services.cms.impl.Utils.isPersonalPublicFolder(node)
+                                        || org.exoplatform.services.cms.impl.Utils.isPersonalPublicFolder(node)) + "' ");
```
Comment (boubaker) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/UIDocumentInfo.java]: Isn't it the same:

`   org.exoplatform.services.cms.impl.Utils.isPersonalPublicFolder(node)`
AND
`|| org.exoplatform.services.cms.impl.Utils.isPersonalPublicFolder(node)`

File: core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java [hunk 8]
```diff
@@ -969,4 +973,39 @@ public static boolean isFolder(Node node) throws RepositoryException  {
...
+   }
+   return false;
+ }
+
+  /**
+   *  check if a node is the personal default public folder "Public".
+   * @param node node to check
+   * @return personal default public folder or not
+   */
+ public static boolean isPersonalPublicFolder(Node node) throws Exception {
```
Comment (boubaker) [inline core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java]: In all cases, we use `isPersonalDefaultFolder || isPersonalPublicFolder`, so we should combine them into one single method IMO `isPersonalDefaultFolder`

File: apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js (hunk 6 above)
Comment (wkhessairi) [reply apps/portlet-explorer/src/main/webapp/javascript/eXo/ecm/UIListView.js]: I just did it like other actions to be consistent and readable.

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/UIDocumentInfo.java (hunk 7 above)
Comment (wkhessairi) [reply core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/UIDocumentInfo.java]: oups it is a typo, I fix it

File: core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java (hunk 8 above)
Comment (wkhessairi) [reply core/services/src/main/java/org/exoplatform/services/cms/impl/Utils.java]: Okay, I just divided into two methods to factorize code and for further usage

**Pull request merged**